{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNW6oXqm3YR2QZykETC9jyP",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/sou-06/AlignedFasta_To_Excel/blob/main/Aligned_Fasta_to_Excel.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "â† Upload **aligned fasta** file from folder ğŸ“ icon. Drag and drop your file.\n",
        "\n",
        "â†“ Parameter setting and input file name. Select color from (https://matplotlib.org/2.0.2/examples/color/named_colors.html)\n"
      ],
      "metadata": {
        "id": "eQYjqFZWpyKm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "inputfile_name = \"your_input_file.fasta\" #Please change here for your uploaded filename.\n",
        "outputfile = \"output.xlsx\"\n",
        "\n",
        "default_bgcolor = \"\"\n",
        "Color_threshold = [0.50, 0.75]\n",
        "bg_colors = [\"lightgray\", \"black\"]\n",
        "font_colors = [\"black\", \"white\"]\n",
        "\n",
        "Consensus_threshold = [0.50, 0.75] # first for lowercase, second for uppercase.\n",
        "row_break_length = 120\n",
        "font = \"Courier New\"\n",
        "\n",
        "start_res = 0     #If you want to output a portion, change this. Specify the position of the first gene's residue.\n",
        "end_res   = 10000\n"
      ],
      "metadata": {
        "id": "8LEnD545oXpI"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "inputfile_name = \"your_input_file.fasta\" #Please change here for your uploaded filename.\n",
        "outputfile = \"output.xlsx\"\n",
        "\n",
        "Color_threshold = [0.50, 0.75]\n",
        "bg_colors = [\"lightgray\", \"black\"]\n",
        "font_colors = [\"black\", \"white\"]\n",
        "\n",
        "Consensus_threshold = [0.50, 0.75] # first for lowercase, second for uppercase.\n",
        "row_break_length = 120\n",
        "font = \"Courier New\"\n",
        "\n",
        "start_res = 30     #If you want to output a portion, change this. Specify the position of the first gene's residue.\n",
        "end_res   = 90\n"
      ],
      "metadata": {
        "id": "2CoMTvu2xJ5w"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "_IxrMF56W_rU",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c17458e5-357d-423e-e220-8e7ad31e6ee0"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: openpyxl in /usr/local/lib/python3.12/dist-packages (3.1.5)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.12/dist-packages (from openpyxl) (2.0.0)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[?25l   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m0.0/3.2 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m\u001b[91mâ•¸\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m161.3 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m82.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.86\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython openpyxl"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "import matplotlib\n",
        "from openpyxl import Workbook\n",
        "from openpyxl.styles import Alignment, PatternFill, Font\n",
        "from openpyxl.utils import get_column_letter\n",
        "from collections import Counter"
      ],
      "metadata": {
        "id": "ZRuUQIAEZpce"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Import\n",
        "ids = []\n",
        "sequences = []\n",
        "for record in SeqIO.parse(inputfile_name, \"fasta\"):\n",
        "    ids.append(record.id)\n",
        "    sequences.append(str(record.seq).upper())\n",
        "\n",
        "# Length adjustment\n",
        "max_len = max(len(seq) for seq in sequences)\n",
        "aligned_seqs = [seq.ljust(max_len, \"-\") for seq in sequences]\n",
        "\n",
        "# Making Excel\n",
        "wb = Workbook()\n",
        "ws = wb.active\n",
        "\n",
        "# Counta for everygene\n",
        "pos_counter = [0] * len(ids)\n",
        "\n",
        "ws.cell(row=1, column=1, value=\"ID\")\n",
        "\n",
        "block_idx = 0  # Block number\n",
        "\n",
        "\n",
        "# For partial output\n",
        "res_map = []\n",
        "res_counter = 0\n",
        "if start_res == 0:\n",
        "    res_map.append(0)\n",
        "\n",
        "for i, base in enumerate(aligned_seqs[0]):\n",
        "    if base != \"-\":\n",
        "        res_counter += 1\n",
        "        if start_res <= res_counter <= end_res:\n",
        "            res_map.append(i)\n",
        "\n",
        "if res_counter < end_res:\n",
        "    res_map.append(max_len - 1)\n",
        "\n",
        "start_idx = min(res_map)\n",
        "end_idx   = max(res_map) + 1\n",
        "\n",
        "for i, seq in enumerate(sequences):\n",
        "    for j in range(start_idx):\n",
        "        if seq[j] != \"-\":\n",
        "            pos_counter[i] += 1\n",
        "\n",
        "for block_start in range(start_idx, end_idx, row_break_length):\n",
        "    block_end = min(block_start + row_break_length, end_idx)\n",
        "    block_idx += 1\n",
        "\n",
        "    # First line of block\n",
        "    base_row = (block_idx - 1) * (len(ids) + 3) + 1\n",
        "\n",
        "    # Res_Number\n",
        "    for row_idx, seq_id in enumerate(ids, start=base_row + 1):\n",
        "        cell = ws.cell(row = row_idx, column = 1, value = seq_id)\n",
        "    # consensus\n",
        "    consensus_row = base_row + len(ids) + 1\n",
        "    ws.cell(row=consensus_row, column=1, value=\"Consensus\")\n",
        "\n",
        "    # == Prosess each block ==\n",
        "    for col_idx in range(block_start, block_end):\n",
        "        excel_col = col_idx - block_start + 2\n",
        "\n",
        "        # --- Header ---\n",
        "        column_label = \"\"\n",
        "        if aligned_seqs[0][col_idx] != \"-\":\n",
        "            count = pos_counter[0] + 1\n",
        "            if count % 10 == 0:\n",
        "                cell = ws.cell(row=base_row, column=excel_col, value = str(count))\n",
        "                #cell.alignment = Alignment(shrinkToFit=True)\n",
        "\n",
        "        # --- homology ---\n",
        "        column_bases = [seq[col_idx] for seq in aligned_seqs]\n",
        "        bases = [b for b in column_bases if b not in [\"-\", \"*\"]]\n",
        "\n",
        "        if len(bases) == 0:\n",
        "            homology = 0\n",
        "            most_common_base = \"-\"\n",
        "        else:\n",
        "            cnt = Counter(bases)\n",
        "            most_common_base, val = cnt.most_common(1)[0]\n",
        "            homology = val / len(column_bases)\n",
        "\n",
        "\n",
        "        for j, seq in enumerate(aligned_seqs):\n",
        "            # Write residue\n",
        "            cell = ws.cell(row=base_row + j + 1, column=excel_col, value=seq[col_idx])\n",
        "            cell.font = Font(name=font)\n",
        "            # Count\n",
        "            if seq[col_idx] not in [\"-\", \"*\"]:\n",
        "                pos_counter[j] += 1\n",
        "\n",
        "            # Coloring\n",
        "            t = len(Color_threshold) - 1\n",
        "            if default_bgcolor != \"\":\n",
        "                cell.fill = PatternFill(\"solid\", fgColor = matplotlib.colors.cnames[default_bgcolor][1:])\n",
        "            while t >= 0 and seq[col_idx] == most_common_base:\n",
        "                if homology >= Color_threshold[t]:\n",
        "                    fill_color = matplotlib.colors.cnames[bg_colors[t]][1:]\n",
        "                    font_color = matplotlib.colors.cnames[font_colors[t]][1:]\n",
        "                    cell.fill = PatternFill(\"solid\", fgColor = fill_color)\n",
        "                    cell.font = Font(color=font_color, name=font)\n",
        "                    break\n",
        "                t -= 1\n",
        "\n",
        "        # --- Consensus ---\n",
        "        if homology >= Consensus_threshold[1]:\n",
        "            ws.cell(row=consensus_row, column=excel_col, value=most_common_base.upper())\n",
        "        elif homology >= Consensus_threshold[0]:\n",
        "            ws.cell(row=consensus_row, column=excel_col, value=most_common_base.lower())\n",
        "\n",
        "        # --- Col width ---\n",
        "        col_letter = get_column_letter(excel_col)\n",
        "        ws.column_dimensions[col_letter].width = 2\n",
        "\n",
        "    # --- Regidue number at right side ---\n",
        "    right_col = min(row_break_length + 2, end_idx - start_idx + 2)\n",
        "    for j in range(len(ids)):\n",
        "        cell = ws.cell(row=base_row + j + 1, column=right_col, value = pos_counter[j])\n",
        "\n",
        "    ws.column_dimensions[get_column_letter(right_col)].width = 6\n",
        "\n",
        "# Change font family\n",
        "for row in ws.iter_rows():\n",
        "    for cell in row:\n",
        "        cell.font = Font(name=font ,color=cell.font.color)\n",
        "\n",
        "wb.save(outputfile)\n",
        "print(\"Excel exported:\", outputfile)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fHCstKd2makS",
        "outputId": "b87b55cb-0884-4c24-cb86-592c3022d8db"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Excel exported: output.xlsx\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "170Mt-quuN-N"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}